src/box/lua/slab.cc

```diff
@@ -265,6 +275,126 @@ lbox_slab_check(MAYBE_UNUSED struct lua_State *L)
 	return 0;
 }
 
+#if defined(HAVE_MALLOC_INFO)
+/*
+ * Decodes and returns the XML document returned by malloc_info() as is.
+ * This is an internal function that isn't supposed to be called by users,
+ * but it may be useful for developers.
+ */
+static int
+lbox_malloc_internal_info(struct lua_State *L)
+{
+	char *buf;
+	size_t buf_size;
+	FILE *fp = open_memstream(&buf, &buf_size);
+	if (fp == NULL) {
+		diag_set(SystemError, "failed to open memory stream");
+		return luaT_error(L);
+	}
+	if (malloc_info(0, fp) != 0) {
+		diag_set(SystemError, "failed to get malloc info");
+		return luaT_error(L);
+	}
+	fclose(fp);
+	lua_pushlstring(L, buf, buf_size);
+	free(buf);
+	return luaT_xml_decode(L);
+}
+
+/*
+ * Returns the malloc memory usage information in a table
+ *
+ *   {
+ *     size = <total allocated>,
+ *     used = <actually used>,
+ *   }
+ *
+ * (all numbers are in bytes).
+ */
+static int
+lbox_malloc_info(struct lua_State *L)
+{
+	lbox_malloc_internal_info(L);
+	/*
+	 * The XML document name returned by malloc_info is expected to be
+	 * "malloc" so the document content should be in malloc[1].
+	 */
+	assert(lua_istable(L, -1));
+	lua_getfield(L, -1, "malloc");
+	if (lua_istable(L, -1))
+		lua_rawgeti(L, -1, 1);
```

Comment:
AFAIU from the comment, you expect `{malloc = {<..our element..>}}`. Why there is no error otherwise?

Reply:
Or, well, other variant of an error reporting (I'm not sure which one is the best for a statistics reporting endpoint).

Reply:
If the `malloc` table is absent, then the version won't be set and we'll fail below. I rewritten to make the check more obvious. Also, instead of an error we now return `{size = 0, used = 0}` as agreed.

<details><summary>diff</summary>

```diff
diff --git a/src/box/lua/slab.cc b/src/box/lua/slab.cc
index fc833516f9b4..0006a59f02d1 100644
--- a/src/box/lua/slab.cc
+++ b/src/box/lua/slab.cc
@@ -310,10 +317,16 @@ lbox_malloc_internal_info(struct lua_State *L)
  *   }
  *
  * (all numbers are in bytes).
+ *
+ * The information is retrieved with malloc_info(). If it isn't supported by
+ * the system or its format is unknown, {size = 0, used = 0} is returned.
  */
 static int
 lbox_malloc_info(struct lua_State *L)
 {
+	int version = 0;
+	uint64_t total = 0;
+	uint64_t available = 0;
 	lbox_malloc_internal_info(L);
 	/*
 	 * The XML document name returned by malloc_info is expected to be
@@ -321,33 +334,31 @@ lbox_malloc_info(struct lua_State *L)
 	 */
 	assert(lua_istable(L, -1));
 	lua_getfield(L, -1, "malloc");
-	if (lua_istable(L, -1))
-		lua_rawgeti(L, -1, 1);
+	if (!lua_istable(L, -1))
+		goto out;
+	lua_rawgeti(L, -1, 1);
+	if (!lua_istable(L, -1))
+		goto out;
 	/*
 	 * First, check the malloc_info version. The only known version is 1.
 	 * It's pointless to proceed if the version is different.
 	 */
-	int version = 0;
-	if (lua_istable(L, -1)) {
-		lua_getfield(L, -1, "version");
-		version = lua_tonumber(L, -1);
-		lua_pop(L, 1);
-	}
+	lua_getfield(L, -1, "version");
+	version = lua_tonumber(L, -1);
+	lua_pop(L, 1);
 	if (version != 1)
-		return luaL_error(L, "unknown malloc info format");
+		goto out;
 	/*
 	 * Extract the size of used memory. Even though the document version is
 	 * valid, we still need to be careful accessing it.
 	 */
-	uint64_t total = 0;
-	uint64_t available = 0;
 	assert(lua_istable(L, -1));
 	lua_getfield(L, -1, "system");
 	if (lua_istable(L, -1)) {
@@ -385,6 +396,7 @@ lbox_malloc_info(struct lua_State *L)
 		}
 	}
 	lua_pop(L, 1);
+out:
 	/* Return memory usage information. */
 	lua_newtable(L);
 	luaL_pushuint64(L, total);
```

</details>